openmc/mgxs/mgxs.py

```diff
@@ -5891,3 +5895,448 @@ def get_units(self, xs_type='macro'):
+        meshsurface_filter = openmc.MeshSurfaceFilter(mesh)
+        filters = [[meshsurface_filter, energy_filter]]
+
+        return self._add_angle_filters(filters)
+
+    @property
+    def xs_tally(self):
+        if self._xs_tally is None:
+            if self.tallies is None:
+                msg = 'Unable to get xs_tally since tallies have ' \
+                      'not been loaded from a statepoint'
+                raise ValueError(msg)
+
+            self._xs_tally = self.rxn_rate_tally
+            self._compute_xs()
+
+        return self._xs_tally
+
+    def load_from_statepoint(self, statepoint):
+        """Extracts tallies in an OpenMC StatePoint with the data needed to
+        compute multi-group cross sections.
+        This method is needed to compute cross section data from tallies
+        in an OpenMC StatePoint object.
+        NOTE: The statepoint must first be linked with an OpenMC Summary object.
+        Parameters
+        ----------
+        statepoint : openmc.StatePoint
+            An OpenMC StatePoint object with tally data
+        Raises
+        ------
+        ValueError
+            When this method is called with a statepoint that has not been
+            linked with a summary object.
+        """
+
+        cv.check_type('statepoint', statepoint, openmc.statepoint.StatePoint)
+
+        if statepoint.summary is None:
+            msg = 'Unable to load data from a statepoint which has not been ' \
+                  'linked with a summary file'
+            raise ValueError(msg)
+
+        # Override the domain object that loaded from an OpenMC summary file
+        # NOTE: This is necessary for micro cross-sections which require
+        # the isotopic number densities as computed by OpenMC
+        geom = statepoint.summary.geometry
+        if self.domain_type in ('cell', 'distribcell'):
+            self.domain = geom.get_all_cells()[self.domain.id]
+        elif self.domain_type == 'universe':
+            self.domain = geom.get_all_universes()[self.domain.id]
+        elif self.domain_type == 'material':
+            self.domain = geom.get_all_materials()[self.domain.id]
+        elif self.domain_type == 'mesh':
+            self.domain = statepoint.meshes[self.domain.id]
+        else:
+            msg = 'Unable to load data from a statepoint for domain type {0} ' \
+                  'which is not yet supported'.format(self.domain_type)
+            raise ValueError(msg)
+
+        # Use tally "slicing" to ensure that tallies correspond to our domain
+        # NOTE: This is important if tally merging was used
+        if self.domain_type == 'mesh':
+            filters = []
+            filter_bins = []
+        elif self.domain_type != 'distribcell':
+            filters = [_DOMAIN_TO_FILTER[self.domain_type]]
+            filter_bins = [(self.domain.id,)]
+        # Distribcell filters only accept single cell - neglect it when slicing
+        else:
```

Comment:
this case is explicitly forbidden by the previous loop

Reply:
Can you please elaborate? The distribcell domain are allowed in the previous loop.

Reply:
My bad, didn't see.